File: src/main/scala/calculator.scala
import scala.math._
import org.scala_tools.time.Imports._

class Calculator {
  val gramsInUnit = 12D
  def alcoholInGramsInServing(serving: Serving) = serving.units * gramsInUnit

  def weightInKilos(weight: Double) = weight / 1000
  def hourInMillis = 3600 * 1000
  def burnRateInMillis(weight: Double) = 0.1 * weightInKilos(weight) / hourInMillis

  def remainingAmount(weight: Double, startGrams: Double, start: DateTime, end: DateTime) = 
    max(startGrams - burnRateInMillis(weight) * (start to end).millis, 0)

  def genderFactor(gender: String) = if (gender == "M") 0.75 else 0.66
  def gramsToBac(user: User, grams: Double) = grams / (genderFactor(user.gender) * user.weight) * 1000

  def calculateBacHistory(user: User, now: DateTime, servings: List[Serving]): List[(Double, DateTime)] = {
    val startTime = if(servings.size > 0) servings.head.date else now
    val start = (0D, startTime)
    val end = Serving(None, None, now, "", 0, 0)
    val gramHistory = (servings ++ List(end)).scanLeft(start)( (last, next) => {
      (remainingAmount(user.weight, last._1, last._2, next.date) + alcoholInGramsInServing(next), next.date)
    })
    gramHistory.map( x => { (gramsToBac(user, x._1), x._2) })
  }
  
}

object Calculator extends Calculator

File: src/server/calculator.ts
import { DateTime } from 'luxon';
import { Serving, User } from './domain';

export const gramsInUnit = 12;
export const hourInMillis = 3600 * 1000;

const alcoholInGramsInServing = (serving: Serving) =>
  serving.units * gramsInUnit;

export const weightInKilos = (weight: number) => weight / 1000;

const genderFactor = (gender: String) => (gender === 'M' ? 0.75 : 0.66);

const burnRateInMillis = (weight: number) =>
  (0.1 * weightInKilos(weight)) / hourInMillis;

const max = (a: number, b: number) => (a > b ? a : b);

const remainingAmount = (
  weight: number,
  startGrams: number,
  start: DateTime,
  end: DateTime,
) =>
  max(startGrams - burnRateInMillis(weight) * end.diff(start).milliseconds, 0);

const gramsToBac = (user: User, grams: number) =>
  (grams / (genderFactor(user.gender) * user.weight)) * 1000;

const formatBac = (num: number) => (Math.round(num * 100) / 100).toFixed(2);

export type GramsAt = {
  at: DateTime;
  grams: number;
};

const calculateBacHistory = (
  user: User,
  now: DateTime,
  servings: Serving[],
) => {
  const endServing: Serving = {
    id: undefined,
    date: now,
    userId: '',
    type: '',
    amount: 0,
    units: 0,
  };
  const servingsUntilNow = [...servings, endServing];
  const startTime = servingsUntilNow[0].date;
  const startEntry: GramsAt = { at: startTime, grams: 0 };

  const gramHistory = servingsUntilNow.reduce(
    (bacHistory: GramsAt[], next: Serving) => {
      const lastEntry = bacHistory[bacHistory.length - 1];
      const nextEntry: GramsAt = {
        grams:
          remainingAmount(
            user.weight,
            lastEntry.grams,
            lastEntry.at,
            next.date,
          ) + alcoholInGramsInServing(next),
        at: next.date,
      };
      return [...bacHistory, nextEntry];
    },
    [startEntry],
  );
  return gramHistory.map((entry) => {
    return gramsToBac(user, entry.grams);
  });
};

export const bacNow = (user: User, servings: Serving[]) =>
  formatBac(calculateBacHistory(user, DateTime.utc(), servings).pop()!);
